### PR TITLE
Use AJAX to update the list of voters dynamically

### PR DESCRIPTION
**Pull request by VSEphpbb**
When Voting up, down or removing your vote, this PR fixes the list of users who voted to dynamically update the list to reflect the correct information.

- [x] Still need to fix username urls. After the user list is updated by AJAX, the username profile links become broken, mainly because of a phpbb_root_path issue. Worst case solution is to not make the username's clickable links to their profiles.

### CODE REVIEW

**Comment by VSEphpbb** [comment]
My fix for the profile URLs having a bad relative path after being updated with AJAX was to force them to use an absolute URL. But if anybody has any better ideas or solutions (or understands why they go bad with AJAX), please speak up.

**Comment by michaelcullum** [comment]
@callumacrae

File: styles/prosilver/template/ideas.js [hunk 1]
```diff
@@ -27,24 +27,27 @@
...
 			});
-			$obj.successVoted.text(message.message)
+			$obj.successVoted.text(result.message)
 				.show()
 				.delay(2000)
 				.fadeOut(300, function() {
 					$obj.votes.fadeIn(300);
-				});
+				})
+			;
```
Comment (callumacrae) [inline styles/prosilver/template/ideas.js]: what

File: styles/prosilver/template/ideas.js [hunk 2]
```diff
@@ -57,7 +60,8 @@
 			.delay(2000)
 			.fadeOut(300, function() {
 				$obj.votes.fadeIn(300);
-			});
+			})
+		;
```
Comment (callumacrae) [inline styles/prosilver/template/ideas.js]: why are you doing this?!

File: styles/prosilver/template/ideas.js [hunk 3]
```diff
@@ -73,13 +77,17 @@
...
+			resetVoteButtons($this);
+			$obj.voteRemove.show();
 		}).fail(voteFailure);
 	});
 
 	$obj.votes.on('click', function(e) {
 		e.preventDefault();
 
-		$obj.votesList.slideToggle();
+		if ($obj.votesList.attr('data-display') === 'true') {
```
Comment (callumacrae) [inline styles/prosilver/template/ideas.js]: `if ($obj.votesList.data('display')) {` would be shorter

File: styles/prosilver/template/ideas.js [hunk 4]
```diff
@@ -326,9 +335,48 @@
...
+		}
+
+		var hasUpVotes = upVoters.length > 0,
+			hasDownVotes = downVoters.length > 0;
+
+		$('#up-voters')
+			.toggle(hasUpVotes)
+			.find('span')
+			.html(upVoters.join(', '))
+		;
```
Comment (callumacrae) [inline styles/prosilver/template/ideas.js]: STOP DOING THIS WHY WOULD YOU DO THIS

File: styles/prosilver/template/ideas.js [hunk 5]
```diff
@@ -326,9 +335,48 @@
...
+
+		$obj.votesList
+			.attr('data-display', (hasUpVotes || hasDownVotes))
+			.toggle(($obj.votesList.is(':visible') && (hasUpVotes || hasDownVotes)))
+		;
+	}
+
+	function resetVoteButtons($this) {
+		$obj.voteUp.add($obj.voteDown).removeClass('dead');
+		if ($this !== undefined) {
```
Comment (callumacrae) [inline styles/prosilver/template/ideas.js]: `if ($this) {` is fine

File: styles/prosilver/template/ideas.js (hunk 2 above)
Comment (VSEphpbb) [inline styles/prosilver/template/ideas.js]: oops, the IDE must have done that

**Comment by callumacrae** [comment]
Why would anyone name anything `$obj` :frowning: 

Even `$elements` is better than that :frowning:

**Comment by VSEphpbb** [comment]
@callumacrae Well, it's an object. Besides, `$elements` is too wordy, and I wanted to keep it short and simple.

**Comment by michaelcullum** [comment]
@VSEphpbb Do you assign objects in php to `$obj`?

**Comment by VSEphpbb** [comment]
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Working_with_Objects#Using_object_initializers

**Comment by VSEphpbb** [comment]
i could rename it `$elem` or `$els`

**Comment by callumacrae** [comment]
Example code probably isn't the best example of good variable names! It is literally just an example object, and so is descriptive. `$obj` is a bad name for an object containing elements, just like `ary` is a bad name for an array containing types of food.

**Comment by VSEphpbb** [comment]
@rxu can you try this out?

**Comment by rxu** [comment]
Sure, will do it later tiday.

**Comment by rxu** [comment]
Tested locally. Works fine for me.

File: event/listener.php [hunk 6]
```diff
@@ -202,15 +202,23 @@ public function show_idea($event)
...
+			$s_voted = false;
+
 			$votes = $this->ideas->get_voters($idea['idea_id']);
 
 			foreach ($votes as $vote)
 			{
 				$this->template->assign_block_vars('votes_' . ($vote['vote_value'] ? 'up' : 'down'), array(
-					'USER'	=> get_username_string('full', $vote['user_id'], $vote['username'], $vote['user_colour']),
-					'S_VOTED' => $this->user->data['user_id'] == $vote['user_id'],
+					'USER' => $vote['user']
```
Comment (Elsensee) [inline event/listener.php]: Missing comma ;)

File: factory/ideas.php [hunk 7]
```diff
@@ -730,4 +746,23 @@ public function get_idea_count()
...
+	/**
+	 * Helper to generate the user profile URL with an
+	 * absolute URL, which helps avoid problems when
+	 * used in AJAX requests.
+	 *
+	 * @return string User profile URL
+	 */
+	protected function profile_url_helper()
+	{
+		static $url;
```
Comment (Elsensee) [inline factory/ideas.php]: I don't know about this. I would add a property for this. Thought static variables in functions are not cool anymore.

**Comment by rxu** [comment]
Tried it again, working. Although when you remove your vote which is the last vote for idea, the rank of 0 points is indicated and when you click it, an empty bar (div) is sliding down.
But I guess that is minor/not related to this PR.

**Comment by michaelcullum** [comment]
@phpbb/frontend

**Comment by VSEphpbb** [comment]
@michaelcullum guess there is no one who cares in @phpbb/frontend

**Comment by michaelcullum** [comment]
That's very annoying.
